File: server/datasets/python_source-code/Multiple-E2/type3_arraysim_1_valid_parentheses.py
def isValid(s):
    fake_stack = [''] * len(s)
    top = -1 
    mapping = {")": "(", "}": "{", "]": "["}
    
    for char in s:
        if char in mapping:
            if top >= 0:
                top_element = fake_stack[top]
                top -= 1
            else:
                top_element = '#'
                
            if mapping[char] != top_element:
                return False
        else:
            top += 1
            fake_stack[top] = char
            
    return top == -1

File: server/datasets/controlled_benchmarks/java/multiple_files/multiple_e2/type3_arraysim_1_valid_parentheses.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public boolean isValid(String s) {
        char[] fake_stack = new char[s.length()];
        int top = -1;
        Map<Character, Character> mapping = new HashMap<>();
        mapping.put(')', '(');
        mapping.put('}', '{');
        mapping.put(']', '[');
        
        for (char ch : s.toCharArray()) {
            if (mapping.containsKey(ch)) {
                char topElement = '#';
                if (top >= 0) {
                    topElement = fake_stack[top];
                    top--;
                }
                
                if (mapping.get(ch) != topElement) {
                    return false;
                }
            } else {
                top++;
                fake_stack[top] = ch;
            }
        }
        return top == -1;
    }
}
